File: Framework/Java/framework/framework/smartkernel/framework/SmartCalendar.java
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
package smartkernel.framework;

import java.util.*;

/**
 * 日历
 * 
 */
public class SmartCalendar {
	/**
	 * 获得一项
	 * 
	 * @param dateTime
	 *            ，日期
	 * @return，结果
	 */
	public static SmartCalendarItem getItem(Date dateTime) {
		Date dataTime = SmartDateTime.parse(SmartDateTime.format(dateTime,
				"yyyy-MM-dd"));

		Calendar calendar = Calendar.getInstance();
		calendar.setTime(dataTime);
		int year = calendar.get(Calendar.YEAR);
		int month = calendar.get(Calendar.MONTH) + 1;
		int weekOfYear = calendar.get(Calendar.WEEK_OF_YEAR);
		int dayOfWeek = calendar.get(Calendar.DAY_OF_WEEK);
		int yearQuarter = 0;
		int dayOfQuarter = 0;

		if (month == 1 || month == 2 || month == 3) {
			yearQuarter = Integer.parseInt(year + "01");
			int totalDays = (int) SmartDateTime.getTimeSpan(
					SmartDateTimeUnitType.Day, dataTime,
					SmartDateTime.parse(SmartDateTime.format(dataTime,
							"yyyy-01-01")));
			dayOfQuarter = totalDays + 1;
		} else if (month == 4 || month == 5 || month == 6) {
			yearQuarter = Integer.parseInt(year + "02");
			int totalDays = (int) SmartDateTime.getTimeSpan(
					SmartDateTimeUnitType.Day, dataTime,
					SmartDateTime.parse(SmartDateTime.format(dataTime,
							"yyyy-04-01")));
			dayOfQuarter = totalDays + 1;
		} else if (month == 7 || month == 8 || month == 9) {
			yearQuarter = Integer.parseInt(year + "03");
			int totalDays = (int) SmartDateTime.getTimeSpan(
					SmartDateTimeUnitType.Day, dataTime,
					SmartDateTime.parse(SmartDateTime.format(dataTime,
							"yyyy-07-01")));
			dayOfQuarter = totalDays + 1;
		} else {
			yearQuarter = Integer.parseInt(year + "04");
			int totalDays = (int) SmartDateTime.getTimeSpan(
					SmartDateTimeUnitType.Day, dataTime,
					SmartDateTime.parse(SmartDateTime.format(dataTime,
							"yyyy-10-01")));
			dayOfQuarter = totalDays + 1;
		}

		SmartCalendarItem result = new SmartCalendarItem();
		result.setDate(dataTime);
		result.setYear(year);
		result.setYearMonth(Integer.parseInt(SmartDateTime.format(dateTime,
				"yyyyMM")));
		result.setYearMonthDay(Integer.parseInt(SmartDateTime.format(dateTime,
				"yyyyMMdd")));
		result.setYearWeek(Integer.parseInt(year
				+ String.format("%02d", weekOfYear)));
		result.setYearWeekDay(Integer.parseInt(year
				+ String.format("%02d", weekOfYear)
				+ String.format("%02d", dayOfWeek)));
		result.setYearQuarter(yearQuarter);
		result.setYearQuarterDay(Integer.parseInt(yearQuarter
				+ String.format("%02d", dayOfQuarter)));

		return result;
	}
}

File: Framework/CSharp/Framework/Framework/SmartCalendar.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using System;
using System.Globalization;

namespace Smartkernel.Framework
{
    /// <summary>
    /// 日历
    /// </summary>
    public static class SmartCalendar
    {
        /// <summary>
        /// 获得一项
        /// </summary>
        /// <param name="dateTime">日期</param>
        /// <returns>结果</returns>
        public static SmartCalendarItem GetItem(DateTime dateTime)
        {
            var dataTime = DateTime.Parse(dateTime.ToString("yyyy-MM-dd"));

            var yearQuarter = 0;
            var dayOfQuarter = 0;

            if (dataTime.Month == 1 || dataTime.Month == 2 || dataTime.Month == 3)
            {
                yearQuarter = Convert.ToInt32(dataTime.Year.ToString() + "01");
                dayOfQuarter = Convert.ToInt32((dataTime - DateTime.Parse(dataTime.ToString("yyyy-01-01"))).TotalDays + 1);
            }
            else if (dataTime.Month == 4 || dataTime.Month == 5 || dataTime.Month == 6)
            {
                yearQuarter = Convert.ToInt32(dataTime.Year.ToString() + "02");
                dayOfQuarter = Convert.ToInt32((dataTime - DateTime.Parse(dataTime.ToString("yyyy-04-01"))).TotalDays + 1);
            }
            else if (dataTime.Month == 7 || dataTime.Month == 8 || dataTime.Month == 9)
            {
                yearQuarter = Convert.ToInt32(dataTime.Year.ToString() + "03");
                dayOfQuarter = Convert.ToInt32((dataTime - DateTime.Parse(dataTime.ToString("yyyy-07-01"))).TotalDays + 1);
            }
            else
            {
                yearQuarter = Convert.ToInt32(dataTime.Year.ToString() + "04");
                dayOfQuarter = Convert.ToInt32((dataTime - DateTime.Parse(dataTime.ToString("yyyy-10-01"))).TotalDays + 1);
            }

            var gregorianCalendar = new GregorianCalendar();

            var result = new SmartCalendarItem
            {
                Date = dataTime,
                Year = dataTime.Year,
                YearMonth = Convert.ToInt32(dataTime.ToString("yyyyMM")),
                YearMonthDay = Convert.ToInt32(dataTime.ToString("yyyyMMdd")),
                YearWeek = Convert.ToInt32(dateTime.Year.ToString() + gregorianCalendar.GetWeekOfYear(dateTime, CalendarWeekRule.FirstDay, DayOfWeek.Sunday).ToString("00")),
                YearWeekDay = Convert.ToInt32(dateTime.Year.ToString() + gregorianCalendar.GetWeekOfYear(dateTime, CalendarWeekRule.FirstDay, DayOfWeek.Sunday).ToString("00") + (Convert.ToInt32(gregorianCalendar.GetDayOfWeek(dataTime)) + 1).ToString("00")),
                YearQuarter = yearQuarter,
                YearQuarterDay = Convert.ToInt32(yearQuarter.ToString() + dayOfQuarter.ToString("00"))
            };

            return result;
        }
    }
}
